src/main/java/one/util/streamex/DoubleWithFirstSpliterator.java

```diff
@@ -0,0 +1,174 @@
+    private DoubleWithFirstSpliterator prefix;
+    private volatile double first;
+    private volatile int state = STATE_NONE;
+    private final DoubleBinaryOperator mapper;
+    private DoubleConsumer action;
+    
+    DoubleWithFirstSpliterator(Spliterator.OfDouble source, DoubleBinaryOperator mapper) {
+        this.source = source;
+        this.mapper = mapper;
+    }
+    
+    private void acquire() {
+        if(lock != null && state == STATE_NONE) {
+            lock.lock();
+        }
+    }
+    
+    private void release() {
+        if(lock != null && lock.isHeldByCurrentThread()) {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean tryAdvance(DoubleConsumer action) {
+        if (state == STATE_NONE) {
+            acquire();
+            try {
+                doInit();
+            }
+            finally {
+                release();
+            }
+        }
+        if (state == STATE_FIRST_READ) {
+            state = STATE_INIT;
+            action.accept(mapper.applyAsDouble(first, first));
+            return true;
+        }
+        if (state != STATE_INIT)
+            return false;
+        this.action = action;
+        boolean hasNext = source.tryAdvance(this);
+        this.action = null;
+        return hasNext;
+    }
+
+    private void doInit() {
+        double prefixState = state;
+        if (prefixState != STATE_NONE)
+            return;
+        if (prefix != null) {
+            prefix.doInit();
+            prefixState = prefix.state;
+        }
+        if (prefixState == STATE_FIRST_READ || prefixState == STATE_INIT) {
+            first = prefix.first;
+            state = STATE_INIT;
+            return;
+        }
+        DoubleConsumer setter = x -> first = x;
+        state = source.tryAdvance(setter) ? STATE_FIRST_READ : STATE_EMPTY;
+    }
+
+    @Override
+    public void forEachRemaining(DoubleConsumer action) {
+        acquire();
+        double myState = state;
+        this.action = action;
+        if(myState == STATE_FIRST_READ || myState == STATE_INIT) {
+            release();
+            if(myState == STATE_FIRST_READ) {
+                state = STATE_INIT;
+                accept(first);
+            }
+            source.forEachRemaining(this);
+            this.action = null;
+            return;
+        }
+        try {
+            DoubleConsumer init = x -> {
+                if (state == STATE_NONE) {
+                    if (prefix != null) {
+                        prefix.doInit();
+                    }
+                    this.first = (prefix == null || prefix.state == STATE_EMPTY) ? x : prefix.first;
+                    state = STATE_INIT;
+                }
+                release();
+            };
+            source.forEachRemaining(init.andThen(this));
+            this.action = null;
+        }
+        finally {
+            release();
+        }
+    }
+
+    @Override
+    public Spliterator.OfDouble trySplit() {

```

Comment:
Ideally the whole `trySplit()` method implementation should be shared in parent abstract class.